orchard_note_encryption.py

```diff
@@ -0,0 +1,301 @@
+        cip = ChaCha20Poly1305(key)
+        return bytes(cip.encrypt(b'\x00' * 12, plaintext))
+
+    @staticmethod
+    def decrypt(key, ciphertext):
+        cip = ChaCha20Poly1305(key)
+        return bytes(cip.decrypt(b'\x00' * 12, ciphertext))
+
+# https://zips.z.cash/protocol/nu5.pdf#saplingandorchardencrypt
+class OrchardNoteEncryption(object):
+    def __init__(self, rand):
+        self._rand = rand
+
+    def encrypt(self, note: OrchardNote, memo, pk_d_new, g_d_new, cv_new, cm_new, ovk=None):
+        np = note.note_plaintext(memo)
+        esk = OrchardKeyAgreement.esk(np.rseed, note.rho)
+        p_enc = bytes(np)
+
+        epk = OrchardKeyAgreement.derive_public(esk, g_d_new)
+        ephemeral_key = bytes(epk)
+        shared_secret = OrchardKeyAgreement.agree(esk, pk_d_new)
+        k_enc = kdf_orchard(shared_secret, ephemeral_key)
+        c_enc = OrchardSym.encrypt(k_enc, p_enc)
+
+        if ovk is None:
+            ock = OrchardSym.k(self._rand)
+            op = self._rand.b(64)
+        else:
+            cv = bytes(cv_new)
+            cmx = bytes(cm_new.extract())
+            ock = prf_ock_orchard(ovk, cv, cmx, ephemeral_key)
+            op = bytes(pk_d_new) + bytes(esk)
+
+        c_out = OrchardSym.encrypt(ock, op)
+
+        self.esk = esk
+        self.shared_secret = shared_secret
+        self.k_enc = k_enc
+        self.p_enc = p_enc
+        self.ock = ock
+        self.op = op
+
+        return TransmittedNoteCipherText(
+            epk, c_enc, c_out
+        )
+
+class TransmittedNoteCipherText(object):
+    def __init__(self, epk, c_enc, c_out):
+        self.epk = epk
+        self.c_enc = c_enc
+        self.c_out = c_out
+
+    def decrypt_using_ivk(self, ivk: Scalar, rho, cm_star):
+        epk = self.epk
+        if epk is None:
+            return None
+
+        shared_secret = OrchardKeyAgreement.agree(ivk, epk)
+        # The protocol spec says to take `ephemeral_key` as input to decryption
+        # and to decode epk from it. That is required for consensus compatibility
+        # in Sapling decryption before ZIP 216, but the reverse is okay here
+        # because Pallas points have no non-canonical encodings.
+        ephemeral_key = bytes(epk)
+        k_enc = kdf_orchard(shared_secret, ephemeral_key)
+        p_enc = OrchardSym.decrypt(k_enc, self.c_enc)
+        if p_enc is None:
+            return None
+
+        leadbyte = p_enc[0]
+        assert(leadbyte == 2)
+        np = OrchardNotePlaintext(
+            p_enc[1:12],   # d
+            Scalar.from_bytes(p_enc[12:20]),  # v
+            p_enc[20:52],  # rseed
+            p_enc[52:564], # memo
+        )
+
+        g_d = diversify_hash(np.d)
+        pk_d = OrchardKeyAgreement.derive_public(ivk, g_d)
+        note = OrchardNote(np.d, pk_d, np.v.s, rho, np.rseed)
+
+        esk = OrchardKeyAgreement.esk(np.rseed, rho)
+        if OrchardKeyAgreement.derive_public(esk, g_d) != epk:
+            return None
```

Comment:
Nit: the spec checks these in reverse:
```suggestion
        esk = OrchardKeyAgreement.esk(np.rseed, rho)
        if OrchardKeyAgreement.derive_public(esk, g_d) != epk:
            return None

        pk_d = OrchardKeyAgreement.derive_public(ivk, g_d)
        note = OrchardNote(np.d, pk_d, np.v.s, rho, np.rseed)
```